### PR TITLE
Add AvalonST basic support

### PR DESCRIPTION
**Pull request by kleinai**
Avalon Streaming has special timing requirements and allowances which make it incompatible with Spinal streams. This PR will provide a way to model and adapt Avalon-ST between IP and Spinal streams.

### CODE REVIEW

**Comment by Dolu1990** [comment]
Cool ^^
What is errorDescriptor: Array[String] = Array(), ?

File: lib/src/main/scala/spinal/lib/bus/avalon/AvalonST.scala [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+                          emptyWithinPacket: Boolean = false,
+                          errorDescriptor: Array[String] = Array(),
+                          symbolEndianness: Endianness = LITTLE,
+                          readyLatency: Int = 0,
+                          readyAllowance: Int = 0) {
+  assert(!(useReady || useValid), "AvalonST must have at least ready OR valid enabled!")
+}
+
+case class AvalonSTPayload(config: AvalonSTConfig) extends Bundle {
+  val data: Bits    = if (config.useData) Bits(config.dataWidth bit) else null
```
Comment (Dolu1990) [inline lib/src/main/scala/spinal/lib/bus/avalon/AvalonST.scala:31]: An shorter way is : 

```scala
val data = config.useData generate Bits(config.dataWidth bits) //:)
```

File: lib/src/main/scala/spinal/lib/bus/avalon/AvalonST.scala [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+  val sop: Bool     = if (config.useSOP) Bool() else null
+}
+
+case class AvalonST(config: AvalonSTConfig) extends Bundle with IMasterSlave {
+
+  val ready: Bool = if (config.useReady) Bool() else null
+  val valid: Bool = if (config.useValid) Bool() else null
+  val payload: AvalonSTPayload = AvalonSTPayload(config)
+
+  lazy val latencyDelay: Bool = Delay(this.ready, config.readyLatency)
```
Comment (Dolu1990) [inline lib/src/main/scala/spinal/lib/bus/avalon/AvalonST.scala:45]: Hmm lazy val in a bundle, i'm realy not sure SpinalHDL will like it.
The goal of those signals is just to cache some values if used right ? Not realy to carry them along the bundle right ?

Comment (kleinai) [reply lib/src/main/scala/spinal/lib/bus/avalon/AvalonST.scala:45]: Missed this. Yeah those are internal for ready tracking. Avalon as some weird timing relationships. I can un-lazy them if you think there might be problems.
Completely forgot up to now that I was working in a Bundle. XP

**Comment by kleinai** [comment]
I think this is ready.  I don't personally use Avalon streams but Intel IP prefers Avalon so I think this is a nice to have as a start.

File: lib/src/main/scala/spinal/lib/bus/avalon/AvalonST.scala (hunk 2 above)
Comment (Dolu1990) [reply lib/src/main/scala/spinal/lib/bus/avalon/AvalonST.scala:45]: Hi ^^
There is still some lazy val, can you un-lazy them ?
If you want, there is the "signalCache" API.

Comment (kleinai) [reply lib/src/main/scala/spinal/lib/bus/avalon/AvalonST.scala]: Finally done. Let me know if I'm using signalCache right.

**Comment by Dolu1990** [comment]
Good to me ^^
you can merge when you feel it.

**Comment by kleinai** [comment]
Will do. Thanks!

**Pull request merged**